### PR TITLE
Fixes for Python 3.8

### PR DESCRIPTION
**Pull request by jspricke**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
     'entry_points': {
         'console_scripts': ['rosversion=rospkg.rosversion:main'],
     },
-    'install_requires': ['catkin_pkg', 'PyYAML'],
+    'install_requires': ['catkin_pkg', 'PyYAML', 'distro'],
```
Comment (dirk-thomas) [inline setup.py]: Nitpick: please insert in alphabetical order.

Please also add the dependencies for the Debian package in the `stdeb.cfg` file.

File: stdeb.cfg [hunk 2]
```diff
@@ -15,8 +15,8 @@ X-Python3-Version: >= 3.4
 Setup-Env-Vars: SKIP_PYTHON_MODULES=1
 
 [rospkg_modules]
-Depends: python-catkin-pkg-modules, python-yaml, lsb-release
-Depends3: python3-catkin-pkg-modules, python3-yaml, lsb-release
+Depends: python-catkin-pkg-modules, python-distro, python-yaml, lsb-release
+Depends3: python3-catkin-pkg-modules, python3-distro, python3-yaml, lsb-release
```
Comment (dirk-thomas) [inline stdeb.cfg]: I could only find the Debian packages for Ubuntu Bionic and newer:
* https://packages.ubuntu.com/search?keywords=python-distro
* https://packages.ubuntu.com/search?keywords=python3-distro

This package needs to target all non-EOL platforms that atm go back until Ubuntu Xenial / Debian Jessie.

**Comment by jspricke** [comment]
This should be fine now, can you have another look?

**Comment by dirk-thomas** [comment]
The fact that changing the import order breaks the test is a problem because that means the behavior of `rospkg` is changing with Python 3.8. We need to make sure that the way it changes (returning `ubuntu` rather than `Ubuntu`) doesn't cause issues downstream. Eventually that means we have to patch code using the API to work with both responses.

**Comment by jspricke** [comment]
I'm not sure what you are talking about, on Ubuntu 18 I get `Ubuntu` with the distro module. I assume that you are seeing some artifact of pip installing the distro module on Ubuntu 16. As this will only be needed for Ubuntu 20, I swapped the import logic on purpose.

**Comment by dirk-thomas** [comment]
I tried the `distro` module on Ubuntu 18.04 and it returned lower case `ubuntu` instead of `Ubuntu`.

**Comment by jspricke** [comment]
```
>>> import distro
>>> distro.linux_distribution()
(u'Ubuntu', u'18.04', u'Bionic Beaver')
```
with python-distro 1.0.1-2
what did you try?

**Comment by dirk-thomas** [comment]
I use the latest version of `distro` installed via `pip`: 1.4.0

And run the command (the same as used in this patch): `distro.linux_distribution(full_distribution_name=0)` and it returns `(u'ubuntu', u'18.04', u'bionic')`.

**Comment by jspricke** [comment]
looks like that's unstable anyhow, for example in Debian I get:
```
>>> import platform
>>> platform.linux_distribution(full_distribution_name=0)
('debian', 'bullseye/sid', '')
```

**Comment by dirk-thomas** [comment]
I split out the Python 3.8 specific fixes for the new SyntaxWarnings and added Python 3.7/3.8 CI: see #183 

I am going to rebase this one after that PR has been merged.

**Comment by dirk-thomas** [comment]
It took me a bit of debugging (a93fa324a37eacf646ac480b86df5516a9379fcc) but as it turns out the inverted import logic exposed that the `import distro` line was never doing what it was supposed to (in Python 2). It was loading the relative module within `rospkg`. See 3fa713bbca526defdb4c3edef3238f2045b43026 for the fix.

**Comment by dirk-thomas** [comment]
I also replaced `ModuleNotFoundError` with `ImportError` (ed980c97907b9dedf85466f28ce50b48fe52e080` ) since the former is only available as of Python 3.6.

**Comment by dirk-thomas** [comment]
With all the back and forth and debugging I felt the need to rewrite one more time. Beside the original rebased commits I only applied two necessary fixes:

* force absolute imports since the rospkg package also contains a module with the name `distro`: 017549e597581d0f14bc12d11f60e8ca6747752b
* ModuleNotFoundError is only available as of Python 3.6: 94ded0aafa2d032e0c3630b0c169eb9275ab44f1

I don't see any evidence of a case mismatch on Travis so I won't change the matching logic from case sensitive to insensitive for now.

@jspricke It would be great if you could take another look and comment if that looks good to you, too.

**Comment by jspricke** [comment]
@dirk-thomas afaik your two commits change nothing as the platform module is part of Python up to 3.8, so the import will always prefer the platform module when it's available. Starting with 3.8 it will switch to the distro module, but in 3.8 `absolute_import` is the default as well as the `ModuleNotFoundError` is available. So LGTM.

**Comment by dirk-thomas** [comment]
Thanks.

**Comment by dirk-thomas** [comment]
This caused #185.

**Comment by jacobperron** [comment]
AFAICT, this is causing CI to fail here: https://github.com/ros-infrastructure/rosdep/pull/735

It appears the [platform module](https://docs.python.org/3.8/library/platform.html) **is** available in Python 3.8. But, `linux_distribution` and `dist` have been removed, so `lsb_info` is always `None`:

https://github.com/ros-infrastructure/rospkg/blob/2c027b17c76ad1a376107c0bbf4b36600d3d572f/src/rospkg/os_detect.py#L140-L145

**Comment by jacobperron** [comment]
Also, `rospkg` tests are failing locally for me with Python 3.8:

<details>
<summary>Test output</summary>

```

$ # nosetests --with-coverage --cover-package=rospkg --with-xunit test
.EE........../root/rospkg/src/rospkg/distro.py:195: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  raw_data = yaml.load(f.read())
.../root/rospkg/test/test_rospkg_distro.py:318: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  assert yaml.load(open(p)) == distro.raw_data, distro.raw_data
./root/rospkg/test/test_rospkg_distro.py:337: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  assert yaml.load(open(p)) == distro.raw_data, distro.raw_data
./root/rospkg/test/test_rospkg_distro.py:353: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  raw_data = yaml.load("""variants:
................................E...................................................
 ======================================================================
ERROR: test.test_rospkg_catkin_packages.test_get_manifest
 ----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/root/rospkg/test/test_rospkg_catkin_packages.py", line 58, in test_get_manifest
    manif = manager.get_manifest("foo")
  File "/root/rospkg/src/rospkg/rospack.py", line 167, in get_manifest
    return self._load_manifest(name)
  File "/root/rospkg/src/rospkg/rospack.py", line 211, in _load_manifest
    retval = self._manifests[name] = parse_manifest_file(self.get_path(name), self._manifest_name, rospack=self)
  File "/root/rospkg/src/rospkg/manifest.py", line 414, in parse_manifest_file
    _static_rosdep_view = init_rospack_interface()
  File "/root/rosdep/src/rosdep2/rospack.py", line 59, in init_rospack_interface
    lookup = _get_default_RosdepLookup(Options())
  File "/root/rosdep/src/rosdep2/main.py", line 132, in _get_default_RosdepLookup
    sources_loader = SourcesListLoader.create_default(sources_cache_dir=options.sources_cache_dir,
  File "/root/rosdep/src/rosdep2/sources_list.py", line 581, in create_default
    matcher = DataSourceMatcher.create_default(os_override=os_override)
  File "/root/rosdep/src/rosdep2/sources_list.py", line 293, in create_default
    os_name, os_version, os_codename = os_detect.detect_os()
  File "/root/rospkg/src/rospkg/os_detect.py", line 664, in detect_os
    raise OsNotDetected("Could not detect OS, tried %s" % attempted)
rospkg.os_detect.OsNotDetected: Could not detect OS, tried ['zorin', 'windows', 'nixos', 'clearlinux', 'ubuntu', 'slackware', 'rhel', 'qnx', 'osx', 'tizen', 'opensuse', 'opensuse', 'opensuse', 'opensuse', 'openembedded', 'neon', 'mx', 'mint', 'linaro', 'gentoo', 'funtoo', 'freebsd', 'fedora', 'elementary', 'elementary', 'debian', 'cygwin', 'centos', 'manjaro', 'arch', 'alpine']

======================================================================
ERROR: test.test_rospkg_catkin_packages.test_licenses
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/root/rospkg/test/test_rospkg_catkin_packages.py", line 65, in test_licenses
    manif = rospack.get_manifest("foo")
  File "/root/rospkg/src/rospkg/rospack.py", line 167, in get_manifest
    return self._load_manifest(name)
  File "/root/rospkg/src/rospkg/rospack.py", line 211, in _load_manifest
    retval = self._manifests[name] = parse_manifest_file(self.get_path(name), self._manifest_name, rospack=self)
  File "/root/rospkg/src/rospkg/manifest.py", line 414, in parse_manifest_file
    _static_rosdep_view = init_rospack_interface()
  File "/root/rosdep/src/rosdep2/rospack.py", line 59, in init_rospack_interface
    lookup = _get_default_RosdepLookup(Options())
  File "/root/rosdep/src/rosdep2/main.py", line 132, in _get_default_RosdepLookup
    sources_loader = SourcesListLoader.create_default(sources_cache_dir=options.sources_cache_dir,
  File "/root/rosdep/src/rosdep2/sources_list.py", line 581, in create_default
    matcher = DataSourceMatcher.create_default(os_override=os_override)
  File "/root/rosdep/src/rosdep2/sources_list.py", line 293, in create_default
    os_name, os_version, os_codename = os_detect.detect_os()
  File "/root/rospkg/src/rospkg/os_detect.py", line 664, in detect_os
    raise OsNotDetected("Could not detect OS, tried %s" % attempted)
rospkg.os_detect.OsNotDetected: Could not detect OS, tried ['zorin', 'windows', 'nixos', 'clearlinux', 'ubuntu', 'slackware', 'rhel', 'qnx', 'osx', 'tizen', 'opensuse', 'opensuse', 'opensuse', 'opensuse', 'openembedded', 'neon', 'mx', 'mint', 'linaro', 'gentoo', 'funtoo', 'freebsd', 'fedora', 'elementary', 'elementary', 'debian', 'cygwin', 'centos', 'manjaro', 'arch', 'alpine']

======================================================================
ERROR: test.test_rospkg_os_detect.test_osx
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/root/rospkg/test/test_rospkg_os_detect.py", line 169, in test_osx
    assert detect.get_codename() == 'snow'
  File "/root/rospkg/src/rospkg/os_detect.py", line 378, in get_codename
    ver = distutils.version.StrictVersion(version).version
AttributeError: 'StrictVersion' object has no attribute 'version'

Name                    Stmts   Miss  Cover
-------------------------------------------
rospkg/__init__.py          7      0   100%
rospkg/common.py           15      0   100%
rospkg/distro.py          316     10    97%
rospkg/environment.py      68      2    97%
rospkg/manifest.py        259     46    82%
rospkg/os_detect.py       469     87    81%
rospkg/rospack.py         223     13    94%
rospkg/stack.py           123     22    82%
-------------------------------------------
TOTAL                    1480    180    88%
----------------------------------------------------------------------
Ran 102 tests in 0.466s

FAILED (errors=3)
```
</details>

**Comment by dirk-thomas** [comment]
> this is causing CI to fail here: ros-infrastructure/rosdep#735

Have you seen the comment above ("This caused #185.") as well as the related fix from #186?

> `rospkg` tests are failing locally for me with Python 3.8:

Do you have the Python package `distro` installed? CI passed with all supported Python version for #186 which pretty much reflects what is on `master` / in the latest release.

**Comment by jacobperron** [comment]
Yes, `distro` was installed (I ran the commands in the travis script).
It's possible I installed `rosdep` before `rospkg` from source and something strange happened. I tried with a fresh docker container and only one test error remains:

```
ERROR: test.test_rospkg_os_detect.test_osx
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/rospkg/test/test_rospkg_os_detect.py", line 169, in test_osx
    assert detect.get_codename() == 'snow'
  File "/rospkg/src/rospkg/os_detect.py", line 378, in get_codename
    ver = distutils.version.StrictVersion(version).version
AttributeError: 'StrictVersion' object has no attribute 'version'
```
I'm not on osx, so maybe it's expected?

---

In any case, if I install  `rospkg` at `master`, I can't seem to confirm the distro on Ubuntu Focal:

```
>>> from rospkg.os_detect import LsbDetect
>>> foo = LsbDetect("ubuntu")
>>> foo.is_os()
False
```

Upper case "Ubuntu" doesn't work either. Here's the [Dockerfile](https://gist.githubusercontent.com/jacobperron/bc990204d796f42b3a26d183ecfe5808/raw/5641c0e1bbdf4e453b4437535cfb1ee108243d47/Dockerfile) I used to produce.

**Comment by sloretz** [comment]
> I'm not on osx, so maybe it's expected?

`StrictVersion` only has a `version` attribute if the [version string has a `patch` component](https://github.com/python/cpython/blob/cbfafa3e3625dae96ce392c88c793f8af55167bf/Lib/distutils/version.py#L142-L145). The [OSX detector gets the version string by executing a file](https://github.com/ros-infrastructure/rospkg/blob/fc3f2b86ea7e3eb37c5bb65bdc0f8adf87d3e884/src/rospkg/os_detect.py#L386). That file has a shebang expects an executable called `python`, but with that `Dockerfile` there is none, causing `stdout` to be an empty string. Installing Python 2 fixes the test failure, but I think we could get rid of that test dependency by making the shebang `/bin/sh` and using `echo`.

https://github.com/ros-infrastructure/rospkg/blob/fc3f2b86ea7e3eb37c5bb65bdc0f8adf87d3e884/test/os_detect/osx/sw_vers#L1-L3

**Pull request merged**